typescript/agentkit/src/action-providers/compound/utils.ts

```diff
@@ -0,0 +1,515 @@
+    const collateralValue = supply.supplyAmount.mul(supply.price);
+    const adjustedValue = collateralValue.mul(supply.collateralFactor);
+    totalAdjustedCollateral = totalAdjustedCollateral.add(adjustedValue);
+  }
+
+  return borrowValue.eq(0) ? new Decimal(Infinity) : totalAdjustedCollateral.div(borrowValue);
+};
+
+/**
+ * Get health ratio after a hypothetical withdraw
+ *
+ * @param wallet - The wallet provider instance
+ * @param cometAddress - The address of the Comet contract
+ * @param tokenAddress - The address of the token contract
+ * @param amount - The amount to withdraw
+ * @returns The health ratio after withdraw as a Decimal
+ */
+export const getHealthRatioAfterWithdraw = async (
+  wallet: EvmWalletProvider,
+  cometAddress: `0x${string}`,
+  tokenAddress: `0x${string}`,
+  amount: bigint,
+): Promise<Decimal> => {
+  if (!tokenAddress || !cometAddress) {
+    throw new Error("Address undefined");
+  }
+
+  const borrowDetails = await getBorrowDetails(wallet, cometAddress);
+  const supplyDetails = await getSupplyDetails(wallet, cometAddress);
+  const borrowValue = borrowDetails.borrowAmount.mul(borrowDetails.price);
+  let totalAdjustedCollateral = new Decimal(0);
+
+  for (const supply of supplyDetails) {
+    const supplyTokenSymbol = supply.tokenSymbol;
+    const withdrawTokenSymbol = await getTokenSymbol(wallet, tokenAddress);
+
+    if (supplyTokenSymbol === withdrawTokenSymbol) {
+      const decimals = await getTokenDecimals(wallet, tokenAddress);
+      const withdrawAmountHuman = new Decimal(formatAmountFromDecimals(amount, decimals));
+      const newSupplyAmount = supply.supplyAmount.sub(withdrawAmountHuman);
+      const assetValue = newSupplyAmount.mul(supply.price);
+      totalAdjustedCollateral = totalAdjustedCollateral.add(
+        assetValue.mul(supply.collateralFactor),
+      );
+    } else {
+      totalAdjustedCollateral = totalAdjustedCollateral.add(
+        supply.supplyAmount.mul(supply.price).mul(supply.collateralFactor),
+      );
+    }
+  }
+
+  return borrowValue.eq(0) ? new Decimal(Infinity) : totalAdjustedCollateral.div(borrowValue);
+};
+
+/**
+ * Get health ratio after a hypothetical borrow
+ *
+ * @param wallet - The wallet provider instance
+ * @param cometAddress - The address of the Comet contract
+ * @param amount - The amount to borrow
+ * @returns The health ratio after borrow as a Decimal
+ */
+export const getHealthRatioAfterBorrow = async (
+  wallet: EvmWalletProvider,
+  cometAddress: `0x${string}`,
+  amount: bigint,
+): Promise<Decimal> => {
+  if (!cometAddress) {
+    throw new Error("Comet address is undefined");
+  }
+
+  const borrowDetails = await getBorrowDetails(wallet, cometAddress);
+  const supplyDetails = await getSupplyDetails(wallet, cometAddress);
+
+  const baseToken = (await wallet.readContract({
+    address: cometAddress,
+    abi: COMET_ABI,
+    functionName: "baseToken",
+    args: [],
+  })) as `0x${string}`;
```

Comment:
We made some improvements to the type arguments on `readContract`, so you may not need this cast anymore!